body_test.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+var allTheBytes []byte = []byte{1, 3, 7, 15, 31, 63, 127, 255}
+
+func TestBodyReading(t *testing.T) {
+
+	filename := "test-files/body-reading"
+
+	_file, _err := os.Create(filename)
+	if _err != nil {
+		t.Error(_err)
+	}
+	for i := 0; i < 256; i++ {
+		_file.Write(allTheBytes)
+	}
+	_file.Close()
+
+	filename2 := "test-files/select-single-byte"
+	_file, _err = os.Create(filename2)
+	if _err != nil {
+		t.Error(_err)
+	}
+
+	_file.Write([]byte{0, 1, 2, 3, 4, 5, 6, 7, 8})
+	_file.Close()
+
+	t.Run("should return error if file doesn't exist", func(t *testing.T) {
+		_, err := newBody("file", 128, 128)
+		if !os.IsNotExist(err) {
+			t.Error(err)
+		}
+	})
+
+	t.Run("should return error if size is zero", func(t *testing.T) {
+		_, err := newBody(filename, 128, 0)
+		if os.IsNotExist(err) {
+			t.Error(err)
+		} else if err == nil {
+			t.Error("Expected an error")
+		}
```

Comment:
I think you could get a nil pointer here if err is nil, maybe more idiomatic would be:

```
if err == nil {
  t.Fatal("Expected an error")
}
if os.IsNotExist(err) {
  t.Fatal(err)
}
```

Reply:
i'm not sure how to interpret this, is the concern about using two `if` instead of `else if` or is it that I'm checking a value of `err` before I've verified that it's non-nil?  Out of curiosity, I tested it and `os.IsNotExist(nil)` returns `false` instead of an error.

Reply:
Yes, both of those really. So taking a step back, the main objective of this review was really just to provide some insight into the types of issues you hit in go programming, rather than only being solely about code correctness (which of course is also important). It is quite idiomatic not to chain `if { ... } else if { .... }` in go, but rather to exit in a code block if you are done processing, and then to have a seperate `if` statement afterwards. I remember reading a go blog about this (I'll see if I can find it). Another common practice is to always check for nil first, if you are going need to check for it. In this case it technically isn't needed, it is just a useful habit to get into, and perhaps makes it less likely that bugs slip in when code is refactored to test for other error conditions later, that might require that `err != nil` etc.

Reply:
cool!  that's really strange to me about the not chained if-elseif-else... good to know.

I totally agree that a nil check should've been first though!

Reply:
So for reference:

* https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow
* [golint](https://github.com/golang/lint) reports non-conformance